`UnitControl`: forward `onBlur` prop as expected

ciampo opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Make the `UnitControl` component call the `onBlur` prop when either the quantity or unit inputs fire the `blur` event.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

While working on [#39522](https://github.com/WordPress/gutenberg/pull/39522/files#diff-c9b6422f332979151b7228eb2af812e84859338e6b4d5ef38e5341ef6e8f5a39L145), I realised that `UnitControl` doesn't forward the `onBlur` prop, despite declaring it in its types.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

The current approach is to call the `onBlur` prop when either the underlying quantity `input` _or_ the unit `select` lose focus.

Alternative approaches would be:
- remove the `onBlur` prop from `UnitControl`'s types
- add 2 separate `blur`-related callbacks, something like `onQuantityBlur` and `onUnitBlur`

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

<details>

<summary>1. Apply this patch</summary>


```diff --git a/packages/components/src/unit-control/stories/index.tsx b/packages/components/src/unit-control/stories/index.tsx
index 87dfd3a84c..2459da2993 100644
--- a/packages/components/src/unit-control/stories/index.tsx
+++ b/packages/components/src/unit-control/stories/index.tsx
@@ -62,6 +62,9 @@ const DefaultTemplate: ComponentStory< typeof UnitControl > = ( {
 					setValue( v );
 					onChange?.( v, extra );
 				} }
+				onBlur={ ( event ) => {
+					console.log( 'Hello', event );
+				} }
 			/>
 		</div>
 	);
```
</details>

2. Open `UnitControl`s Storybook example
3. Move the keyboard focus around, paying particular attention to how the `onBlur` callback gets fired when iteracting with the quantity `input` and the unit `select`

ciampo committed:
Add `onBlur` support to `UnitControl`

ciampo committed:
CHANGELOG

aaronrobertshaw commented on packages/components/src/unit-control/index.tsx:
```diff
@@ -165,7 +166,10 @@ function UnforwardedUnitControl(
 
-	const handleOnBlur: FocusEventHandler< HTMLInputElement > = mayUpdateUnit;
+	const handleOnBlur: FocusEventHandler< HTMLInputElement > = ( event ) => {
+		mayUpdateUnit( event );
```
It looks like the change handlers and setting the unit are only called via `mayUpdateUnit` if `isPressEnterToChange` is `true`. Should we tweak the logic?

ciampo replied:
I thought about this, and I'm not sure it would make any difference:

- changing the `unit` from the select dropdown already fires the `onUnitChange` and `onChange` callbacks
- typing a unit in the `NumberControl` input is only allowed when the `isPressEnterToChange` prop is set to `true` — which is already the logic implemented in `mayUpdateUnit`

Maybe I'm missing something?

stokesman replied:
That's right, `mayUpdateUnit` is only to check if a unit can be parsed from the number input and intentionally limited to `isPressEnterToChange` being `true`.

aaronrobertshaw replied:
Ok, thanks. That makes sense to me 👍

aaronrobertshaw reviewed (commented):
Thanks for this fix @ciampo 👍 

The `onBlur` prop gets called when I would expect it to. Personally, I prefer this PR's option over the alternatives listed in the description.

I had one question regarding the calling of `mayUpdateUnit` within `handleOnBlur` and left that as an inline comment.

Should we also document the `onBlur` prop in the `README.md`?

stokesman reviewed (approved):
This looks good and tests well for me. I agree the chosen approach is better than the alternatives listed in the description. I think it'd be good to move forward with this approach for now.

I also think it warrants a follow up about consistency with our other composite components. For instance, one of the first things I thought to check when reviewing this was how it's handled in `RangeControl`. There, focus and blur are only forwarded from the range input and not the number input. So that one feels a bit lacking as well.

ciampo committed:
Add unit tests

ciampo committed:
README

ciampo commented:
Thank you both for validating the current approach for when to fire the `onBlur` callback.

I went ahead and:

- rebased to solve a conflict
- listed the `onBlur` prop in the README
- added unit tests

> I had one question regarding the calling of `mayUpdateUnit` within `handleOnBlur` and left that as an inline comment.

I replied to that comment, but I'll wait for @aaronrobertshaw's reply to make sure that we're on the same page.

> I also think it warrants a follow up about consistency with our other composite components.

That's a great point. I've opened #39615 to start keeping track of it

stokesman commented on packages/components/src/unit-control/test/index.js:
```diff
@@ -220,6 +220,94 @@ describe( 'UnitControl', () => {
+			);
+		} );
+
+		it( 'should run onBlur, onChange and onUnitChange callbacks when isPressEnterToChange is true and the component is blurred with an uncommitted value', async () => {
```
Great addition since, minus the check for `onBlur`, a test like this should have existed well before this PR. Nit: given that the other added tests cover it, asserting `onBlur` is called here seems redundant, no?

ciampo replied:
Good point, I just pushed a quick update. Will wait for CI checks to be 🟢  and merge!

stokesman reviewed (approved):
The added unit tests are great and I verified they fail without the changes from fe212af0d218c30b6cd4d9ea8f6d5617870f894a.

ciampo committed:
Do not test for `onBlur` callback to be invoked
